sound/soc/sof/pcm.c

```diff
@@ -164,9 +142,13 @@ static int sof_pcm_hw_params(struct snd_soc_component *component,
 	 * Handle repeated calls to hw_params() without free_pcm() in
 	 * between. At least ALSA OSS emulation depends on this.
 	 */
-	ret = sof_pcm_dsp_pcm_free(substream, sdev, spcm);
-	if (ret < 0)
-		return ret;
+	if (spcm->prepared[substream->stream] && pcm_ops->hw_free) {
```

Comment:
nit pick: invert the test, the first patch only makes sense if the second is true.